2023/Q1/artifacts/PoS-pseudocode/PoS-model-redelegation.md

```diff
@@ -272,16 +317,58 @@ func unbond(validator_address, delegator_address, total_amount)
     //this serves to check that there are bonds (in the docs) and that these are greater than the amount we are trying to unbond
     if (sum{amount | <start, amount> in delbonds} >= total_amount) then
       var remain = total_amount
-      var amount_after_slashing = total_amount
+      var amount_after_slashing = 0
       //Iterate over bonds and create unbond
       forall (<start, amount> in delbonds while remain > 0) do
+        //Retrieve the bond's associated redelegation record (if any)
+        var redelegation_record = redelegated_bonds[delegator_address][validator_address][start]
         //Take the minimum between the remainder and the unbond. This is equal to amount if remain > amount and remain otherwise 
         var amount_unbonded = min{amount, remain}
+        var redelegated_amount = 0
+        if redelegation_record != ⊥ then
+          redelegated_amount = redelegation_record.amount
+
+        var amount_unbonded_after_slashing = 0
+        //compute amount after slashing if some of the tokens do not come from redelegation
+        if redelegated_amount < amount_unbonded then 
+          // set of slashes that happened while the bond was contributing to the validator's stake
+          var set_slashes = {s | s in slashes[validator_address] && start <= s.epoch }
+          amount_unbonded_after_slashing = compute_amount_after_slashing(set_slashes, amount_unbonded - redelegated_amount)
+        //compute amount after slashing  of the tokens coming from redelegation
+        if redelegation_record != ⊥ then
+          set_slashes = set_slashes \union {slash | slash in slashes[redelegation_record.validator] &&
+                                                    //start - pipeline_length is the epoch when the redelegation was issued
+                                                    //any slashed processed at an epoch < slash - pipeline_length - unbonding_length has been applied already
+                                                    start - pipeline_length - unbonding_length <= slash.epoch < start}
```

Comment:
> If I understand correctly, for a re-delegation we cannot determine the start epoch on the source validator, so we're just applying the slash indiscriminately as long as it's in the slashable duration here, right?

> I don't think it's feasible to apply these exactly unless we do something even more complicated, as the re-delegation might be created from multiple bond deltas each with different start epoch, so I think it's is a good approximation, but would like to confirm and document this.

We do not need to know when it started because we are slashing when doing the redelegation. Thus, we are not doing any approximation. Let me elaborate a bit:
- When the redelegation is issued, we compute the redelegated amount after slashing the source validator.
- If the redegelation happens at `e`, this means that we have processed all slashes up to `e-unbonding_lengh` (not included): an infraction at an epoch is processed at the end of that epoch + `unbonding_length`.
- Thus, when ubonding, we need to apply slashes for the source validator from `e-unbonding_lengh` to `e+pipeline_lentgh` (not included): when the redelegation ends.

Does this clarify it?

Reply:
Sry, I'm not sure if I follow. Let me try to ask with 2 examples:

1. Say you have a delegation to validator A with start epoch `e`. In epoch `e+1` you redelegate to validator B. After that, there's evidence discovered for validator A from epoch `e-1`. Would the redelegation get slashed for it when its being processed at `e-1+unbonding` epoch? In this case, the original bond didn't contribute to the validator A voting power when the misbehavior occurred.
2. As another example, let's say that the delegation to validator A has start epoch `e`. Now you redelegate to B in `e+2`, after which an evidence for A is found for `e+1`. Does this differ from the first example when the slash is being processed at `e+1+unbonding` epoch? In this case, the original bond did contribute to the voting power of the validator when it misbehaved.

It seems to me that in both cases the slash would be applied to the redelegation.

Reply:
> 1. Say you have a delegation to validator A with start epoch `e`. In epoch `e+1` you redelegate to validator B. After that, there's evidence discovered for validator A from epoch `e-1`. Would the redelegation get slashed for it when its being processed at `e-1+unbonding` epoch? In this case, the original bond didn't contribute to the validator A voting power when the misbehavior occurred.

I see your concern. This is probably a problem: I believe in the current design we would slash. I am not sure there is a way to cut corners and avoid tracking the start of the delegation.

Reply:
I think it’s an okay compromise. For redelegation there will be small chance that you may get slashed when you shouldn’t be, but not the other way around so it’s not a safety issue

Reply:
Brought this up internally and we agreed we're okay with tracking the re-delegations with start epochs of their respective source bonds to make this precise. We're assuming epoch to last about a day and it's not gonna speed up significantly in future because of other concerns, so this shouldn't pose a DoS vector

Reply:
Addressed in [32102ae](https://github.com/informalsystems/partnership-heliax/pull/45/commits/32102aecfb854459ff15d81bc356898fcbba2d04).